File: 0029-divide-two-integers/0029-divide-two-integers.java
class Solution {
    public int divide(int dividend, int divisor) {
        if(dividend==Integer.MIN_VALUE && divisor == -1) return Integer.MAX_VALUE;
        
        boolean positive = (dividend<0 && divisor<0) || (dividend>0 && divisor>0);
        
        Long dividend_long = Long.valueOf(dividend);
        Long divisor_long = Long.valueOf(divisor);
        
        dividend_long = Math.abs(dividend_long);
        divisor_long = Math.abs(divisor_long);
        
        if(dividend_long < divisor_long) return 0;
        List<List<Long>> list = new ArrayList<List<Long>>();
        list.add(new ArrayList<Long>(Arrays.asList(Long.valueOf(divisor_long),Long.valueOf(1))));
        
        while(list.get(list.size()-1).get(0) <= dividend_long) {
            Long first = list.get(list.size()-1).get(0);
            Long second = list.get(list.size()-1).get(1);
            list.add(new ArrayList<Long>(Arrays.asList(first+first, second+second)));
        }
        
        int ans = 0;
        int i=list.size()-1;
        
        while(i>=0 && dividend_long>=divisor_long) {
            if(list.get(i).get(0) <= dividend_long) {
                ans += list.get(i).get(1);
                dividend_long -= list.get(i).get(0);
            }
            i--;
        }
        
        return positive ? ans : -ans;
    }
}

File: 0029-divide-two-integers/0029-divide-two-integers.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        if(dividend == INT_MIN && divisor == -1) return INT_MAX;
        long divid = dividend;
        long divis = divisor;
        
        bool sign = true;
        if(dividend<0 && divisor>0) sign = false;
        if(dividend>0 && divisor<0) sign =false;
        
        if(dividend < 0) divid = dividend*(long)-1;
        if(divisor < 0) divis = divisor*(long)-1;
        
        if(divis>divid) return 0;
        
        long ans = 1;
        long sum = divis;
        
        while(sum<divid) {
            sum = sum*2;
            ans = ans*2;
        }
        
        if(sum==divid) return sign? ans:-1*ans;
        
        long res = ans/2 + divide(divid-sum/2 , divis);
        
        return sign?res:-1*res;
    }
};